File: src/main/java/raylras/zen/langserver/provider/SemanticTokensProvider.java
package raylras.zen.langserver.provider;

import org.eclipse.lsp4j.SemanticTokens;
import org.eclipse.lsp4j.SemanticTokensParams;
import raylras.zen.code.CompilationUnit;
import raylras.zen.code.Listener;
import raylras.zen.langserver.Document;
import raylras.zen.util.Range;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public class SemanticTokensProvider extends Listener {

    public static CompletableFuture<SemanticTokens> semanticTokensFull(Document doc, SemanticTokensParams params) {
        return doc.getUnit().map(unit -> CompletableFuture.supplyAsync(() -> {
            SemanticTokensProvider provider = new SemanticTokensProvider(unit);
            return new SemanticTokens(provider.data);
        })).orElseGet(SemanticTokensProvider::empty);
    }

    public static CompletableFuture<SemanticTokens> empty() {
        return CompletableFuture.completedFuture(null);
    }

    private final CompilationUnit unit;
    private final List<Integer> data = new ArrayList<>();
    private int prevLine = Range.FIRST_LINE;
    private int prevColumn = Range.FIRST_COLUMN;

    private SemanticTokensProvider(CompilationUnit unit) {
        this.unit = unit;
    }

    private void push(Range range, int tokenType, int tokenModifiers) {
        if (range == null) return;
        int line = range.start().line() - prevLine;
        int column = range.start().line() == prevLine ? range.start().column() - prevColumn : range.start().column();
        int length = range.end().column() - range.start().column();
        prevLine = range.start().line();
        prevColumn = range.start().column();
        data.add(line);
        data.add(column);
        data.add(length);
        data.add(tokenType);
        data.add(tokenModifiers);
    }

}

File: zenscript-language-server/src/main/kotlin/raylras/zen/lsp/provider/SemanticTokensProvider.kt
package raylras.zen.lsp.provider

import org.antlr.v4.runtime.tree.ParseTree
import org.eclipse.lsp4j.SemanticTokens
import org.eclipse.lsp4j.SemanticTokensRangeParams
import raylras.zen.lsp.provider.data.TokenModifier
import raylras.zen.lsp.provider.data.TokenType
import raylras.zen.lsp.provider.data.tokenModifier
import raylras.zen.lsp.provider.data.tokenType
import raylras.zen.model.CompilationUnit
import raylras.zen.model.Listener
import raylras.zen.model.parser.ZenScriptParser.*
import raylras.zen.model.resolve.resolveSymbols
import raylras.zen.model.symbol.Symbol
import raylras.zen.util.*

object SemanticTokensProvider {
    fun semanticTokensRange(unit: CompilationUnit, params: SemanticTokensRangeParams): SemanticTokens {
        val listener = SemanticTokensListener(unit, params.range.toTextRange())
        unit.accept(listener)
        return SemanticTokens(listener.result)
    }
}

private class SemanticTokensListener(private val unit: CompilationUnit, private val range: TextRange) : Listener() {
    val result = mutableListOf<Int>()
    private var prevLine: Int = BASE_LINE
    private var prevColumn: Int = BASE_COLUMN

    override fun enterQualifiedName(ctx: QualifiedNameContext) {
        checkRange(ctx) {
            resolveSymbols<Symbol>(ctx, unit).firstOrNull()?.let {
                push(ctx.simpleName().last().textRange, it.tokenType, it.tokenModifier)
            }
        }
    }

    override fun exitVariableDeclaration(ctx: VariableDeclarationContext) {
        checkRange(ctx.simpleName()) { simpleName ->
            unit.symbolMap[simpleName]?.let {
                push(simpleName.textRange, TokenType.VARIABLE, it.tokenModifier + TokenModifier.DECLARATION.bitflag)
            }
        }
    }

    override fun exitFunctionDeclaration(ctx: FunctionDeclarationContext) {
        checkRange(ctx.simpleName()) { simpleName ->
            unit.symbolMap[simpleName]?.let {
                push(simpleName.textRange, TokenType.FUNCTION, it.tokenModifier + TokenModifier.DECLARATION.bitflag)
            }
        }
    }

    override fun exitFormalParameter(ctx: FormalParameterContext) {
        checkRange(ctx.simpleName()) { simpleName ->
            push(
                simpleName.textRange,
                // Workaround: in VSCode, it seems that read-only Parameter are not highlighted as expected, but Variable working fine.
                TokenType.VARIABLE,
                TokenModifier.READONLY.bitflag + TokenModifier.DECLARATION.bitflag
            )
        }
    }

    override fun exitSimpleNameExpr(ctx: SimpleNameExprContext) {
        checkRange(ctx) {
            resolveSymbols<Symbol>(ctx, unit).firstOrNull()?.let {
                push(ctx.textRange, it.tokenType, it.tokenModifier)
            }
        }
    }

    override fun exitMemberAccessExpr(ctx: MemberAccessExprContext) {
        checkRange(ctx.simpleName()) { simpleName ->
            resolveSymbols<Symbol>(ctx, unit).firstOrNull()?.let {
                push(simpleName.textRange, it.tokenType, it.tokenModifier)
            }
        }
    }

    private fun checkRange(ctx: ParseTree, callback: (ParseTree) -> Unit) {
        if (ctx.textRange in range) {
            callback(ctx)
        }
    }

    private fun push(range: TextRange?, tokenType: TokenType?, tokenModifiers: Int) {
        range ?: return
        tokenType ?: return
        val line = range.start.line - prevLine
        val column = if (range.start.line == prevLine) range.start.column - prevColumn else range.start.column
        val length = range.end.column - range.start.column
        prevLine = range.start.line
        prevColumn = range.start.column
        result.add(line)
        result.add(column)
        result.add(length)
        result.add(tokenType.ordinal)
        result.add(tokenModifiers)
    }
}
